build/wrapper.cc

```diff
@@ -180,31 +185,33 @@ ray::core::CoreWorker &_GetCoreWorker() {
 }
 
 // Example of using `CoreWorker::Put`: https://github.com/ray-project/ray/blob/ray-2.5.1/src/ray/core_worker/test/core_worker_test.cc#L224-L237
-ObjectID put(const std::shared_ptr<RayObject> object,
-             const std::vector<ObjectID> &contained_object_ids) {
+Status put(const std::shared_ptr<RayObject> object,
+             const std::vector<ObjectID> &contained_object_ids,
+             ObjectID *object_id) {
 
     auto &worker = CoreWorkerProcess::GetCoreWorker();
-
     // Store our data in the object store
-    ObjectID object_id;
-    RAY_CHECK_OK(worker.Put(*object, contained_object_ids, &object_id));
-    return object_id;
+    auto status = worker.Put(*object, contained_object_ids, object_id);
+    return status;
 }
 
 // Example of using `CoreWorker::Get`: https://github.com/ray-project/ray/blob/ray-2.5.1/src/ray/core_worker/test/core_worker_test.cc#L210-L220
-std::shared_ptr<RayObject> get(const ObjectID object_id, int64_t timeout_ms) {
+Status get(const ObjectID object_id, int64_t timeout_ms, std::vector<std::shared_ptr<RayObject>> *objects) {
```

Comment:
(I agree with @omus suggestion here though)

Reply:
So..having trouble with the "re-instante the vector" part...

We give a pointer to a ` shared_ptr<RayObject>` (`result`), but the coreworker function takes a pointer to a `vector<shared_ptr<RayObject>>` (`result_vec`).

It's not as simple as wrappring `result` in a vector since then we'd have a pointer to a vector of _pointers_.
And it's also not viable to set the `result` to point to `&result_vec[0]` after the fact either since I'm pretty sure that memory is released once `get` returns.

I tried to `memcpy` the actual data in `result_vec` into the address of `result` but that wasn't working out either.

I want to completely avoid `reinterpret_cast` and friends since those are too dangerous but I'm open to any ideas?

IMO, I think passing a `vector<shared_ptr<RayObject>>` would be fine as  it's consistent with the function signature of the coreworker functions. It's also much easier to extract the `only` value in the `results` on the julia-side.

Reply:
> I also tried to memcpy the actual data in result_vec into the address of result but that wasn't working out either.
Specifically I used `std::memcpy(result, &result_vec[0], sizeof(shared_ptr<RayObject>))`

Reply:
> In any case IMO, I think passing a vector<shared_ptr<RayObject>> would be fine as it's consistent with the function signature of the coreworker functions. It's also much easier to extract the only value in the results on the julia-side.

Okay, that's fine.  Don't get too bogged down here.  I'm annoyed that we are leaking so much of the C++ stuff into the Ray.jl codebase but it seems like that ship has sailed, especially since we've lost the distinction between the jll packager and Ray.jl itself.

Just leave a comment in the julia code explaining why you have to wrap in a vector.

Reply:
> And it's also not viable to set the `result` to point to `&result_vec[0]` after the fact either since I'm pretty sure that memory is released once get returns.

As we're dealing with `shared_ptr`s here as long as you have one pointer to the memory you are using it will be retained. We should definitely be able to do this but I suspect you'll need to have a `std::shared_ptr<RayObject> *` so you can update the passed in parameter. I'll note this [`std::swap`](https://cplusplus.com/reference/memory/shared_ptr/swap-free/) function kinda does what we want but I don't think it's necessary

Reply:
This is my impression on how this would be done:
```cpp
Status get(const ObjectID object_id, int64_t timeout_ms, std::shared_ptr<RayObject> *result) {
    std::vector<std::shared_ptr<RayObject>> objects;
    auto status = worker.Get(get_obj_ids, timeout_ms, &objects);
    *result = objects[0];
    ...
}
```
The `std::shared_ptr`'s data is stored on the heap.

Reply:
I will give this a shot..I had initially tried `result = &objects[0]` but I realise in hindsight I was just reassigning the _address_ of `result` and not the contents.

Reply:
worked 👍